### PR TITLE
feat: Add update deadline prompt system

### PR DESCRIPTION
**Pull request by rtylerdavis**
## Summary

Implements the update deadline prompt system proposed in #1130. When `WAU_UpdateDeadlineDays` is configured, WAU presents a WPF dialog to logged-in users listing pending updates with per-app deadlines instead of silently installing. Fully backward-compatible -- zero effect when not configured.

Update Prompt Dialog

## Features

- Per-app deadline tracking (clock starts on first detection, never resets on version bumps)
- WPF dialog via ServiceUI.exe with per-app selective update checkboxes
- Final-day enforcement: apps due today are auto-checked, locked, highlighted in red -- Remind button hidden
- Color-coded urgency (red for today/overdue, amber for 1-3 days, white for 4+)
- Configurable reminder interval (`WAU_ReminderIntervalDays`)
- Custom company branding (`WAU_CompanyName`)
- User-context app support (respects `WAU_UserContext`)
- Whitelist/blacklist respected in deadline tracking
- Self-healing: stale entries purged automatically, failed updates preserved for retry
- X button blocked, auto-dismiss timer

## Configuration

| Setting | Default | Description |
|---------|---------|-------------|
| `WAU_UpdateDeadlineDays` | `0` (disabled) | Days from first detection to forced update |
| `WAU_ReminderIntervalDays` | `2` | Days between reminder prompts |
| `WAU_CompanyName` | *(empty)* | Company name in prompt header |

Dedicated `WAU-Notifier.admx`/`.adml` included for Intune/GPO deployment. Uses a separate ADMX namespace (`WAUNotifier.Policies`) to coexist with the existing `WAU.admx`. The three settings are also added to the upstream `WAU.admx`/`.adml` for single-file deployments.

## Files Changed

**Modified:**
- `Winget-Upgrade.ps1` -- deadline mode if/else branch around main update loop
- `config/WAU-MSI_Actions.ps1` -- registers/unregisters UpdatePrompt and UpdateNow tasks

**New:**
- `WAU-UpdatePrompt.ps1` -- WPF dialog, user interaction, partial update logic
- `WAU-UpdateNow.ps1` -- processes updates from pending-updates.json
- `functions/Get-UpdateDeadlines.ps1` -- reads/purges deadline registry entries
- `functions/Set-UpdateDeadline.ps1` -- creates/updates deadline entries
- `functions/Start-UpdatePromptTask.ps1` -- writes JSON, fires prompt task
- `WAU-Notifier.admx` / `WAU-Notifier.adml` -- ADMX policy definitions

## Testing

Tested on x64 and ARM64 (Snapdragon Surface Laptop) via both manual install and Intune deployment. Full test plan covering install/uninstall lifecycle, deadline creation, prompt interaction, overdue enforcement, user-context apps, whitelist/blacklist, ADMX policy delivery, and edge cases.

---

*Related: #1121*

### CODE REVIEW

**Comment by Romanitho** [comment]
Nice, massive piece of work! I'll need some time to review and approve the code. It won't be part of the next release, but I'll target the one after that.
As mentioned, I’m also starting a new repo for the C# version. This is a new challenge for me, and I’m planning to rewrite the code to keep WAU’s core functionality while improving the overall logic.
Since the project began, many things have evolved and features were added over time; I think the move to C# is the perfect opportunity for a full refactor. I specifically want to use the Winget COM API to make the agent simpler and more robust. This deadline prompt system would, of course, become a central feature of the new WAU agent.

**Comment by Romanitho** [comment]
To maintain brand consistency and avoid any user confusion, would it be possible to use the native WAU icon instead?

**Comment by rtylerdavis** [comment]
> To maintain brand consistency and avoid any user confusion, would it be possible to use the native WAU icon instead?

of course, I just threw that in as a demo to make sure it worked, at first I was just using the info.png from your source, but I'm open whatever icon you'd like!

File: Sources/Winget-AutoUpdate/functions/Get-UpdateDeadlines.ps1 [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        catch {
+            Write-ToLog "WARNING: Unparseable FirstDetected for $appId ('$($props.FirstDetected)')" "Yellow"
+        }
+
+        try {
+            $deadline = [DateTime]::Parse($props.Deadline)
+        }
+        catch {
+            Write-ToLog "WARNING: Unparseable Deadline for $appId ('$($props.Deadline)')" "Yellow"
+        }
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/functions/Get-UpdateDeadlines.ps1:90]: Good catch, switched both DateTime::Parse calls to ParseExact('yyyy-MM-dd', InvariantCulture) to match the format  we write in Set-UpdateDeadline. Fixed in 763bb8b.

File: Sources/Winget-AutoUpdate/Winget-Upgrade.ps1 [hunk 2]
```diff
@@ -269,6 +326,53 @@ if (Test-Network) {
...
+                New-Item -ItemType Directory -Path $sharedDir -Force | Out-Null
+            }
+            try {
+                $acl = Get-Acl $sharedDir
+                $authUsersRule = New-Object System.Security.AccessControl.FileSystemAccessRule(
+                    'Authenticated Users', 'Modify', 'ContainerInherit,ObjectInherit', 'None', 'Allow')
+                if (-not ($acl.Access | Where-Object {
+                    $_.IdentityReference -eq 'NT AUTHORITY\Authenticated Users' -and
+                    $_.FileSystemRights -band [System.Security.AccessControl.FileSystemRights]::Modify
+                })) {
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/Winget-Upgrade.ps1]: Using 'Authenticated Users' breaks on non-English systems. Switched to SID S-1-5-11 for both the rule creation and the existing-rule check, consistent with the SID-based pattern in WAU-MSI_Actions.ps1. Fixed in 763bb8b.

File: Sources/Winget-AutoUpdate/functions/Set-UpdateDeadline.ps1 [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    if (Test-Path $AppRegPath) {
+
+        # Entry exists -- only update AvailableVersion if a newer version is now available.
+        # FirstDetected and Deadline are intentionally preserved (deadline never resets).
+        $existing = Get-ItemProperty -Path $AppRegPath -ErrorAction SilentlyContinue
+        if ($existing.AvailableVersion -ne $App.AvailableVersion) {
+            Set-ItemProperty -Path $AppRegPath -Name "AvailableVersion" -Value $App.AvailableVersion
+            Write-ToLog "Deadline entry updated (new version): $($App.Id) -- $($existing.AvailableVersion) -> $($App.AvailableVersion)"
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/functions/Set-UpdateDeadline.ps1]: The -ne check (any change, not just "newer") is the intended behavior, if winget reports a different version string for any reason we want to record it. Updated the comment to say "update AvailableVersion when it changes." Fixed in 763bb8b.

File: Sources/Winget-AutoUpdate/WAU-UpdatePrompt.ps1 [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+    Exit 1
+}
+
+if (-not $pendingData.Apps -or @($pendingData.Apps).Count -eq 0) {
+    Exit 0
+}
+
+$reminderDays = 2
+if ($pendingData.Config -and $null -ne $pendingData.Config.ReminderIntervalDays) {
+    $reminderDays = [int]$pendingData.Config.ReminderIntervalDays
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/WAU-UpdatePrompt.ps1]: Added TryParse with a minimum of 1, falling back to the default of 2. Fixed in 763bb8b.

File: Sources/Winget-AutoUpdate/WAU-UpdatePrompt.ps1 [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+}
+#endregion READ PENDING UPDATES
+
+#region BUILD ROW OBJECTS
+$today   = (Get-Date).Date
+$appRows = [System.Collections.Generic.List[WauAppRow]]::new()
+
+foreach ($app in @($pendingData.Apps)) {
+    $deadline = $null
+    try { $deadline = [DateTime]::Parse($app.Deadline) } catch { continue }
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/WAU-UpdatePrompt.ps1]: Same fix as the Get-UpdateDeadlines comment - switched to ParseExact('yyyy-MM-dd', InvariantCulture) with a null/whitespace guard. Fixed in 763bb8b.

File: .github/workflows/GA_Close-Inactive-Issues.yml [hunk 6]
```diff
@@ -12,6 +12,7 @@ jobs:
     permissions:
       issues: write
       pull-requests: write
+      contents: write
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply .github/workflows/GA_Close-Inactive-Issues.yml]: This was a stray diff from branching off main instead of develop. Rebased the branch onto develop - this file is no longer in the PR.

File: Sources/Winget-AutoUpdate/WAU-UpdatePrompt.ps1 [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+
+        # Remind for unselected apps
+        $nextPromptTime = (Get-Date).AddDays($reminderDays).ToString('o')
+        try {
+            Set-ItemProperty -Path $WAURegPath -Name 'NextPromptTime' -Value $nextPromptTime
+        }
+        catch { }
+    }
+    else {
+        # Full update -- clear any stale NextPromptTime from a previous partial snooze
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/WAU-UpdatePrompt.ps1]: Good catch. The full-update path now rewrites pending-updates.json from in-memory data before firing UpdateNow, same as the partial-update path already does. This guards against the main SYSTEM cycle overwriting the file while the prompt is open. Fixed in 763bb8b.

File: Sources/Winget-AutoUpdate/Winget-Upgrade.ps1 [hunk 8]
```diff
@@ -297,50 +401,256 @@ if (Test-Network) {
...
-                        Write-ToLog "$($app.Name) : Skipped upgrade because it is *wildcard* in the excluded app list" "Gray"
+                }
+
+                # Step 4: Show deadline prompt for pending apps if a user is logged in and
+                # the snooze window has expired.
+                if ($pendingEntries.Count -gt 0) {
+                    $explorerprocesses = @(Get-CimInstance -Query "SELECT * FROM Win32_Process WHERE Name='explorer.exe'" -ErrorAction SilentlyContinue)
+                    if ($explorerprocesses.Count -eq 0) {
+                        Write-ToLog "No user logged on -- skipping update prompt for $($pendingEntries.Count) pending apps" "Gray"
                     }
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/Winget-Upgrade.ps1:589]: This is intentional. Apps due "today" are kept in the pending/prompt path because the deadline prompt is the user's last chance to see what's about to update before it happens silently. If no user is logged in, the app becomes overdue on the next cycle (at most ~24h depending on the WAU schedule) and gets force-updated then. Changing this to force-update "today" apps when no session is available would mean users who log in later that day never see the prompt, worse UX for the common case to save at most one cycle in an uncommon edge case.

File: Sources/Winget-AutoUpdate/Winget-Upgrade.ps1 [hunk 9]
```diff
@@ -297,50 +401,256 @@ if (Test-Network) {
...
+                $userOutdatedDir = [System.IO.Path]::Combine($env:ProgramData, 'Winget-AutoUpdate')
+                if (-not (Test-Path $userOutdatedDir)) { New-Item -ItemType Directory -Path $userOutdatedDir -Force | Out-Null }
+                $userOutdatedPath = [System.IO.Path]::Combine($userOutdatedDir, 'user-context-outdated.csv')
+                # Clean up old JSON format if present (migrated to CSV)
+                $oldJsonPath = [System.IO.Path]::Combine($userOutdatedDir, 'user-context-outdated.json')
+                if (Test-Path $oldJsonPath) { Remove-Item -Path $oldJsonPath -Force -ErrorAction SilentlyContinue }
+                if ($userEligible.Count -gt 0) {
+                    $userEligible | Select-Object Name, Id, Version, AvailableVersion |
+                        Export-Csv -Path $userOutdatedPath -NoTypeInformation -Encoding UTF8 -Force -ErrorAction Stop
+                    Write-ToLog "$($userEligible.Count) user-context outdated apps written for deadline tracking" "Cyan"
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/Winget-Upgrade.ps1]: Wrapped in try/catch with a log warning. The apps will be picked up on the next cycle. Fixed in 763bb8b.

File: Sources/Winget-AutoUpdate/functions/Start-UpdatePromptTask.ps1 [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        Write-ToLog "ERROR: Could not write pending-updates.json -- $($_.Exception.Message)" "Red"
+        return
+    }
+
+    # Trigger the UpdatePrompt task. This runs WAU-UpdatePrompt.ps1 via ServiceUI.exe
+    # in the logged-in user's desktop session. The main task exits immediately after.
+    $promptTask = Get-ScheduledTask -TaskName "Winget-AutoUpdate-UpdatePrompt" -ErrorAction SilentlyContinue
+    if ($promptTask) {
+        $promptTask | Start-ScheduledTask
+        Write-ToLog "Winget-AutoUpdate-UpdatePrompt task triggered"
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/functions/Start-UpdatePromptTask.ps1]: Wrapped in try/catch with -ErrorAction Stop and a warning log, consistent with the error handling pattern used elsewhere. Fixed in 763bb8b.

**Comment by rtylerdavis** [comment]
> To maintain brand consistency and avoid any user confusion, would it be possible to use the native WAU icon instead?

I grabbed the native icon out of the MSI installer artifact and redistributed under "icons/notify_icon.png" for the WPF dialogue to pickup. Done in 6378d36.

**Comment by Romanitho** [comment]
First of all, thank you for this massive amount of work. This is a highly requested feature and it brings WAU to another level of professionalism.

Given the scope of these changes (new notification system, registry tracking, and scheduled tasks), I am considering bumping WAU to v3.0 with this PR. Since this will be a major version, I'm perfectly fine with having breaking changes. If you feel we need to refactor some parts of the code to make it smoother, simpler, or more robust, please feel free to do so. We have a great opportunity here to clean up the core logic while integrating this new system.

Regarding the ADMX/ADML files: do we really need a separate namespace and separate files?

- Admin Experience: Updating custom ADMX in Intune is already tedious; managing multiple files makes it worse.

- Maintenance: From my side, it's much easier to manage a single All-in-One ADMX within our GitHub Actions. I’d rather not have to adjust the release workflow for multiple files every time a policy changes.

I suggest merging everything into the main ADMX file and incrementing its version number. What do you think about streamlining this?

Thanks again for the great contribution!

**Comment by Romanitho** [comment]
Quick follow-up on the technical side:

I just merged #1135, which introduces some minor conflicts with your PR.
Specifically, `Update-App $app` now supports a custom source via `-src $Script:WingetSourceCustom`.

Sorry for the timing!

**Comment by rtylerdavis** [comment]
> Regarding the ADMX/ADML files: do we really need a separate namespace and separate files?

Definitely not, I had just done that for my unofficial fork/testing, it's an artifact of that. I will get those merged as they should be.

> I just merged #1135, which introduces some minor conflicts with your PR.
> Specifically, `Update-App $app` now supports a custom source via `-src $Script:WingetSourceCustom`.

No worries I'll take a look at this and get it updated today!

File: Sources/Policies/ADMX/WAU.admx [hunk 11]
```diff
@@ -412,5 +412,38 @@
         <text id="UpdatesTimeDelayTime" valueName="WAU_UpdatesTimeDelay" />
       </elements>
     </policy>
+    <policy name="UpdateDeadlineDays_Enable" class="Machine"
+      displayName="$(string.UpdateDeadlineDays_Name)"
+      explainText="$(string.UpdateDeadlineDays_Explain)"
+      key="Software\Policies\Romanitho\Winget-AutoUpdate"
+      presentation="$(presentation.UpdateDeadlineDays)">
+      <parentCategory ref="WAU" />
+      <supportedOn ref="WAU:SUPPORTED_WAU_2_6_0" />
```
Comment (Romanitho) [inline Sources/Policies/ADMX/WAU.admx]: You can go for WAU_3_0_0

File: Sources/Policies/ADMX/WAU.admx [hunk 12]
```diff
@@ -412,5 +412,38 @@
...
+      displayName="$(string.ReminderIntervalDays_Name)"
+      explainText="$(string.ReminderIntervalDays_Explain)"
+      key="Software\Policies\Romanitho\Winget-AutoUpdate"
+      presentation="$(presentation.ReminderIntervalDays)">
+      <parentCategory ref="WAU" />
+      <supportedOn ref="WAU:SUPPORTED_WAU_2_6_0" />
+      <elements>
+        <text id="ReminderIntervalDays" valueName="WAU_ReminderIntervalDays" />
+      </elements>
+    </policy>
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Policies/ADMX/WAU.admx:448]: This is due to merging the admx/adml with upstream, ignoring and resolving.

File: Sources/Policies/ADMX/WAU.admx (hunk 11 above)
Comment (rtylerdavis) [reply Sources/Policies/ADMX/WAU.admx]: Sorry was out last week for a conference, but I'm back. 0870aeb addresses this as well as the copilot issues brought up below!

File: Sources/Winget-AutoUpdate/WAU-UpdatePrompt.ps1 [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+$script:AllowClose = $false
+
+# Block the X button -- users must choose Remind or Update.
+# Button handlers set AllowClose before calling Close().
+$window.Add_Closing({
+    param($eventSender, $e)
+    if (-not $script:AllowClose) {
+        $e.Cancel = $true
+    }
+})
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/WAU-UpdatePrompt.ps1:326]: Added SystemEvents.SessionEnding to AllowClose=$true. Fixed in 0870aeb.

File: Sources/Policies/ADMX/en-US/WAU.adml [hunk 14]
```diff
@@ -153,6 +153,35 @@ This helps prevent many devices from running the task at the exact same time. Th
...
+
+For example, setting this to "Contoso" changes the header to "Contoso requires the following updates to be installed."
+
+If this policy is disabled or not configured, the default text "Your organization" is used.</string>
+      <string id="UpdateDeadlineDays_Name">Update deadline enforcement (days)</string>
+      <string id="UpdateDeadlineDays_Explain">If this policy is enabled, WAU will track pending updates and display a deadline prompt dialog to logged-in users rather than silently installing updates.
+
+The value sets the number of days from first detection until the update is forced silently in the background.
+
+Setting this to 0 (or disabling / not configuring this policy) restores the default silent update behaviour and clears all stored deadline data.
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Policies/ADMX/en-US/WAU.adml]: Fixed in 0870aeb.

File: Sources/Winget-AutoUpdate/Winget-Upgrade.ps1 [hunk 15]
```diff
@@ -297,50 +411,261 @@ if (Test-Network) {
...
+                # Without this, excluded apps (e.g. WAU itself) would get deadline-tracked.
+                if ($UseWhiteList) {
+                    $deadlineApps = @($outdated | Where-Object {
+                        $id = $_.Id
+                        ($toUpdate -contains $id) -or ($toUpdate | Where-Object { $id -like $_ })
+                    })
+                    $skippedCount = $outdated.Count - $deadlineApps.Count
+                    if ($skippedCount -gt 0) {
+                        Write-ToLog "$skippedCount apps excluded from deadline tracking (not in whitelist)" "Gray"
                     }
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/Winget-Upgrade.ps1:485]: Addressed in 0870aeb. Both $skippedCount sites now use @($outdated).Count to handle the scalar return from Get-WingetOutdatedApps when only one app is outdated.

File: Sources/Winget-AutoUpdate/Winget-Upgrade.ps1 [hunk 16]
```diff
@@ -297,50 +411,261 @@ if (Test-Network) {
...
+                }
+                else {
+                    $deadlineApps = @($outdated | Where-Object {
+                        $id = $_.Id
+                        -not ($toSkip -contains $id) -and -not ($toSkip | Where-Object { $id -like $_ })
+                    })
+                    $skippedCount = $outdated.Count - $deadlineApps.Count
+                    if ($skippedCount -gt 0) {
+                        Write-ToLog "$skippedCount apps excluded from deadline tracking (in blacklist)" "Gray"
                     }
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/Winget-Upgrade.ps1:495]: Addressed in 0870aeb. Both $skippedCount sites now use @($outdated).Count to handle the scalar return from Get-WingetOutdatedApps when only one app is outdated.

File: Sources/Winget-AutoUpdate/Winget-Upgrade.ps1 [hunk 17]
```diff
@@ -141,6 +141,63 @@ if (Test-Network) {
...
+                # (they wouldn't be in $deadlineApps), resetting their clocks.
+                $userOutdatedPath = [System.IO.Path]::Combine($env:ProgramData, 'Winget-AutoUpdate', 'user-context-outdated.csv')
+                if (Test-Path $userOutdatedPath) {
+                    $updatedIds = @($userApps | ForEach-Object { $_.Id })
+                    $remaining = @(Import-Csv -Path $userOutdatedPath -Encoding UTF8 |
+                        Where-Object { $_.Id -notin $updatedIds })
+                    if ($remaining.Count -gt 0) {
+                        $remaining | Export-Csv -Path $userOutdatedPath -NoTypeInformation -Encoding UTF8 -Force
+                        Write-ToLog "Removed $($updatedIds.Count) updated apps from user-context-outdated.csv ($($remaining.Count) remaining)"
+                    }
```
[the remark this answers is not in the document]
Comment (rtylerdavis) [reply Sources/Winget-AutoUpdate/Winget-Upgrade.ps1]: Addressed in 0870aeb. The user-context UpdateNow handoff now tracks per-app success via the $Script:InstallOK delta around each Update-App call. Only confirmed-updated app IDs are removed from user-context-outdated.csv, failed updates stay in the CSV so SYSTEM re-merges them next cycle and the deadline registry entry is preserved (no clock reset on transient failure).

**Comment by chiujiewei** [comment]
Would it be possible to add a scheduling option in the prompt window, allowing users to choose install times (e.g., install in 1 hour, 2 hours, etc.) alongside the “Update Now” option?
